pulsar-broker/src/main/java/org/apache/pulsar/PulsarHealthcheck.java

```diff
@@ -0,0 +1,204 @@
+                Boolean.parseBoolean(properties.getProperty("tlsAllowInsecureConnection", "false")));
+        builder.enableTlsHostnameVerification(
+                Boolean.parseBoolean(properties.getProperty("tlsEnableHostnameVerification", "false")));
+        return builder;
+    }
+
+    private static ClientBuilder propsToClientBuilder(Properties properties) throws Exception {
+        ClientBuilder builder = PulsarClient.builder();
+        if (StringUtils.isNotBlank(properties.getProperty("brokerServiceUrl"))) {
+            builder.serviceUrl(properties.getProperty("brokerServiceUrl"));
+        } else {
+            builder.serviceUrl(properties.getProperty("webServiceUrl"));
+        }
+
+        if (StringUtils.isNotBlank(properties.getProperty("authPlugin"))) {
+            builder.authentication(properties.getProperty("authPlugin"), properties.getProperty("authParams"));
+        }
+        builder.tlsTrustCertsFilePath(properties.getProperty("tlsTrustCertsFilePath"));
+        builder.allowTlsInsecureConnection(
+                Boolean.parseBoolean(properties.getProperty("tlsAllowInsecureConnection", "false")));
+        builder.enableTlsHostnameVerification(
+                Boolean.parseBoolean(properties.getProperty("tlsEnableHostnameVerification", "false")));
+        return builder;
+    }
+
+    private static void successExit() {
+        ObjectMapper mapper = new ObjectMapper();
+        Map<String, String> output = new HashMap<>();
+        output.put("result", "ok");
+        try {
+            mapper.writeValue(System.out, output);
+        } catch (IOException ioe) {
+            log.error("Error writing JSON", ioe);
+        }
+        System.exit(0);
+    }
+
+    private static void errorExit(int code, String reason) {
+        ObjectMapper mapper = new ObjectMapper();
+        Map<String, String> output = new HashMap<>();
+        output.put("result", "error");
+        output.put("cause", reason);
+        try {
+            mapper.writeValue(System.out, output);
+        } catch (IOException ioe) {
+            log.error("Error writing JSON", ioe);
+        }
+        System.exit(code);
+    }
+
+    public static void main(String[] args) throws Exception {
+        HealthcheckArguments healthcheckArguments = new HealthcheckArguments();
+        JCommander jcommander = new JCommander(healthcheckArguments);
+        jcommander.setProgramName("PulsarHealthcheck");
+
+        jcommander.parse(args);
+        if (healthcheckArguments.help) {
+            jcommander.usage();
+            System.exit(-1);
+        }
+
+        String configFile = healthcheckArguments.healthcheckConf;
+        Properties properties = new Properties();
+        try (FileInputStream fis = new FileInputStream(configFile)) {
+            properties.load(fis);
+        }
+
+        // no need to clean this executor up, as we exit straight after
+        ExecutorService executor = Executors.newSingleThreadExecutor((r) -> {
```

Comment:
a bit confused here. if we exit straight after, why we need to start an executor? can't we just trigger the callable?

Reply:
To have a timeout on it.

Reply:
hmm, doesn't sound right to me. because both publish and receive have timeouts, we should just rely on that. if you run the callable directly here, you will have a clear stacktrace when timeout happens. putting this in another thread and having timeout on the runnable will hide a lot of useful information when timeout occurs.

Reply:
Admin does not. Admin hangs forever.

Reply:
Admin is essentially a jersey client. We should be able to configure a timeout via `readTimeout` in the jersey client builder. I feel the right approach is to:

- add timeout in pulsar admin. since other applications who use pulsar admin directly can also be benefited from this setting.
- run and call the command directly in the main thread, rather than submitting to another thread, which will make debug easier.

Reply:
I agree that admin should have read and connect timeout exposed, but I didn't go down that route because I thought we needed this sooner. I've created #2891 for this.

Regarding running in the main thread, I added the timeout so that the call to the healthcheck would return in a predictable time. I looked further into k8s docs, and they do provide a timeout option, which defaults to 1 second, which is a problem in itself, as the jvm takes multiple seconds to boot. As such, I think we should move the check into a http endpoint, and recommend that people use a http probe.

Reply:
+1 for http probe

Reply:
One issue to solve around the HTTP probe is permissions. 
We prob need to have super-user access to trigger the health-check, though then the credentials need to be passed by the checker (eg: Kubernetes livenessProbe config). 

The only advantage of using a script in the local machine is that the credentials will be already there.

Reply:
A script on the local machine can also call the http probe. That said, it's not a given that admin credentials are available on the broker. In a securely configured cluster they should not be.
For the probe I'll also add a healthcheckRole configuration option, which will have access to run the probe. There will also need to be configuration options to configure the client for the probe within the endpoint.